### PR TITLE
Add TempoWorld API to Control Actors

### PR DESCRIPTION
**Pull request by pete-tempo**

### CODE REVIEW

File: TempoWorld/Source/TempoWorld/Public/ActorControl.proto [hunk 1]
```diff
@@ -0,0 +1,225 @@
...
+
+message SpawnActorRequest {
+  string type = 1;
+  bool deferred = 2;
+  TempoScripting.Transform transform = 3;
+  string relative_to_actor = 4;
+}
+
+message SpawnActorResponse {
+  string spawned_name = 1;
```
Comment (pete-tempo) [inline TempoWorld/Source/TempoWorld/Public/ActorControl.proto:16]: Add spawned transform

File: TempoWorld/Source/TempoWorld/Public/ActorControl.proto [hunk 2]
```diff
@@ -0,0 +1,225 @@
...
+  string actor = 1;
+  TempoScripting.Transform transform = 2;
+  string relative_to_actor = 3;
+}
+
+message SetComponentTransformRequest {
+  string actor = 1;
+  string component = 2;
+  TempoScripting.Transform transform = 3;
+  bool relative_to_world = 4;
```
Comment (pete-tempo) [inline TempoWorld/Source/TempoWorld/Public/ActorControl.proto:37]: What about root component?

File: TempoWorld/Source/TempoWorld/Public/ActorControl.proto [hunk 3]
```diff
@@ -0,0 +1,225 @@
...
+message FinishSpawningActorRequest {
+  string actor = 1;
+}
+
+message DestroyActorRequest {
+  string actor = 1;
+}
+
+message SetActorTransformRequest {
+  string actor = 1;
```
Comment (pete-tempo) [inline TempoWorld/Source/TempoWorld/Public/ActorControl.proto:28]: Make sure to warn if actor is not specified

File: TempoWorld/Source/TempoWorld/Public/TempoActorControlServiceSubsystem.h [hunk 4]
```diff
@@ -0,0 +1,77 @@
...
+
+	void GetAllComponents(const TempoWorld::GetAllComponentsRequest& Request, const TResponseDelegate<TempoWorld::GetAllComponentsResponse>& ResponseContinuation) const;
+	
+	void GetActorProperties(const TempoWorld::GetActorPropertiesRequest& Request, const TResponseDelegate<TempoWorld::GetPropertiesResponse>& ResponseContinuation) const;
+
+	void GetComponentProperties(const TempoWorld::GetComponentPropertiesRequest& Request, const TResponseDelegate<TempoWorld::GetPropertiesResponse>& ResponseContinuation) const;
+
+protected:
+	template <typename RequestType>
+	void SetProperty(const RequestType& Request, const TResponseDelegate<TempoScripting::Empty>& ResponseContinuation) const;
```
Comment (pete-tempo) [inline TempoWorld/Source/TempoWorld/Public/TempoActorControlServiceSubsystem.h:73]: Just make it private

File: TempoWorld/Source/TempoWorld/Public/TempoWorldUtils.h [hunk 5]
```diff
@@ -2,4 +2,116 @@
...
+		BasePaths.Add(BaseClassPath);
+
+		TSet< FTopLevelAssetPath > Excluded;
+		AssetRegistry.GetDerivedClassNames(BasePaths, Excluded, DerivedNames);
+	}
+	FARFilter Filter;
+	FTopLevelAssetPath BPPath(UBlueprint::StaticClass()->GetPathName());
+	Filter.ClassPaths.Add(BPPath);
+	Filter.bRecursiveClasses = true;
+	// Filter.PackagePaths.Add(TEXT("/Game"));
```
Comment (pete-tempo) [inline TempoWorld/Source/TempoWorld/Public/TempoWorldUtils.h:84]: Remove this

File: TempoWorld/Source/TempoWorld/Private/TempoActorControlServiceSubsystem.cpp [hunk 6]
```diff
@@ -0,0 +1,834 @@
...
+	
+	SpawnActorResponse Response;
+	Response.set_spawned_name(TCHAR_TO_UTF8(*SpawnedActor->GetActorNameOrLabel()));
+
+	ResponseContinuation.ExecuteIfBound(Response, grpc::Status_OK);
+}
+
+void UTempoActorControlServiceSubsystem::FinishSpawningActor(const FinishSpawningActorRequest& Request, const TResponseDelegate<TempoScripting::Empty>& ResponseContinuation)
+{
+	AActor* Actor = GetActorWithName(GetWorld(), UTF8_TO_TCHAR(Request.actor().c_str()));
```
Comment (pete-tempo) [inline TempoWorld/Source/TempoWorld/Private/TempoActorControlServiceSubsystem.cpp:150]: Return early if actor is empty

File: TempoWorld/Source/TempoWorld/Private/TempoActorControlServiceSubsystem.cpp [hunk 7]
```diff
@@ -0,0 +1,834 @@
...
+	Actor->FinishSpawning(*SpawnTransform);
+
+	DeferredSpawnTransforms.Remove(Actor);
+
+	ResponseContinuation.ExecuteIfBound(TempoScripting::Empty(), grpc::Status_OK);
+}
+
+void UTempoActorControlServiceSubsystem::DestroyActor(const TempoWorld::DestroyActorRequest& Request, const TResponseDelegate<TempoScripting::Empty>& ResponseContinuation) const
+{
+	AActor* Actor = GetActorWithName(GetWorld(), UTF8_TO_TCHAR(Request.actor().c_str()));
```
Comment (pete-tempo) [inline TempoWorld/Source/TempoWorld/Private/TempoActorControlServiceSubsystem.cpp:172]: Return early if actor is empty

File: TempoWorld/Source/TempoWorld/Private/TempoActorControlServiceSubsystem.cpp [hunk 8]
```diff
@@ -0,0 +1,834 @@
...
+	}
+	
+	Actor->SetActorTransform(Transform);
+
+	ResponseContinuation.ExecuteIfBound(TempoScripting::Empty(), grpc::Status_OK);
+}
+
+void UTempoActorControlServiceSubsystem::SetComponentTransform(const TempoWorld::SetComponentTransformRequest& Request, const TResponseDelegate<TempoScripting::Empty>& ResponseContinuation) const
+{
+	const AActor* Actor = GetActorWithName(GetWorld(), UTF8_TO_TCHAR(Request.actor().c_str()));
```
Comment (pete-tempo) [inline TempoWorld/Source/TempoWorld/Private/TempoActorControlServiceSubsystem.cpp:212]: Return early if actor or component is empty

File: TempoWorld/Source/TempoWorld/Private/TempoActorControlServiceSubsystem.cpp [hunk 9]
```diff
@@ -0,0 +1,834 @@
...
+	{
+		Component->SetRelativeTransform(Transform);
+	}
+
+	ResponseContinuation.ExecuteIfBound(TempoScripting::Empty(), grpc::Status_OK);
+}
+
+void UTempoActorControlServiceSubsystem::ActivateComponent(const ActivateComponentRequest& Request, const TResponseDelegate<TempoScripting::Empty>& ResponseContinuation) const
+{
+	const AActor* Actor = GetActorWithName(GetWorld(), UTF8_TO_TCHAR(Request.actor().c_str()));
```
Comment (pete-tempo) [inline TempoWorld/Source/TempoWorld/Private/TempoActorControlServiceSubsystem.cpp:242]: Return early if actor or component is empty

File: TempoWorld/Source/TempoWorld/Private/TempoActorControlServiceSubsystem.cpp [hunk 10]
```diff
@@ -0,0 +1,834 @@
...
+	}
+
+	Component->Activate();
+
+	ResponseContinuation.ExecuteIfBound(TempoScripting::Empty(), grpc::Status_OK);
+}
+
+void UTempoActorControlServiceSubsystem::DeactivateComponent(const DeactivateComponentRequest& Request, const TResponseDelegate<TempoScripting::Empty>& ResponseContinuation) const
+{
+	const AActor* Actor = GetActorWithName(GetWorld(), UTF8_TO_TCHAR(Request.actor().c_str()));
```
Comment (pete-tempo) [inline TempoWorld/Source/TempoWorld/Private/TempoActorControlServiceSubsystem.cpp:263]: Return early if actor or component is empty

File: TempoWorld/Source/TempoWorld/Private/TempoActorControlServiceSubsystem.cpp [hunk 11]
```diff
@@ -0,0 +1,834 @@
...
+		ActorDescriptor->set_name(TCHAR_TO_UTF8(*Actor->GetActorNameOrLabel()));
+		ActorDescriptor->set_type(TCHAR_TO_UTF8(*Actor->GetClass()->GetName()));
+	}
+
+	ResponseContinuation.ExecuteIfBound(Response, grpc::Status_OK);
+}
+
+void UTempoActorControlServiceSubsystem::GetAllComponents(const GetAllComponentsRequest& Request, const TResponseDelegate<GetAllComponentsResponse>& ResponseContinuation) const
+{
+	const AActor* Actor = GetActorWithName(GetWorld(), UTF8_TO_TCHAR(Request.actor().c_str()));
```
Comment (pete-tempo) [inline TempoWorld/Source/TempoWorld/Private/TempoActorControlServiceSubsystem.cpp:346]: Return early if actor is empty

File: TempoWorld/Source/TempoWorld/Private/TempoActorControlServiceSubsystem.cpp [hunk 12]
```diff
@@ -0,0 +1,834 @@
...
+	GetAllComponentsResponse Response;
+
+	TArray<UActorComponent*> Components;
+	Actor->GetComponents(Components);
+	for (const UActorComponent* Component : Components)
+	{
+		TempoWorld::ComponentDescriptor* ComponentDescriptor = Response.add_components();
+		ComponentDescriptor->set_name(TCHAR_TO_UTF8(*Component->GetName()));
+		ComponentDescriptor->set_type(TCHAR_TO_UTF8(*Component->GetClass()->GetName()));
+		ComponentDescriptor->set_actor(TCHAR_TO_UTF8(*Actor->GetActorNameOrLabel()));
```
Comment (pete-tempo) [inline TempoWorld/Source/TempoWorld/Private/TempoActorControlServiceSubsystem.cpp:361]: Readme needs to clarify uniqueness of names in different modes

File: TempoWorld/Source/TempoWorld/Private/TempoActorControlServiceSubsystem.cpp [hunk 13]
```diff
@@ -0,0 +1,834 @@
...
+			{
+				Type = TEXT("color");
+				if (Value)
+				{
+					FLinearColor ValueLinearColor;
+					StructProperty->GetValue_InContainer(Object, &ValueLinearColor);
+					const FColor ValueColor = ValueLinearColor.ToFColor(true);
+					*Value = FString::Printf(TEXT("r:%d g:%d b:%d"), ValueColor.R, ValueColor.G, ValueColor.B);
+				}
+			}
```
Comment (pete-tempo) [inline TempoWorld/Source/TempoWorld/Private/TempoActorControlServiceSubsystem.cpp:480]: else unsupported

File: TempoWorld/Source/TempoWorld/Private/TempoActorControlServiceSubsystem.cpp [hunk 14]
```diff
@@ -0,0 +1,834 @@
...
+		}
+		PropertyDescriptor->set_name(TCHAR_TO_UTF8(*Property->GetName()));
+		PropertyDescriptor->set_type(TCHAR_TO_UTF8(*Type));
+		PropertyDescriptor->set_value(TCHAR_TO_UTF8(*Value));
+	}
+}
+
+void UTempoActorControlServiceSubsystem::GetActorProperties(const GetActorPropertiesRequest& Request, const TResponseDelegate<GetPropertiesResponse>& ResponseContinuation) const
+{
+	const AActor* Actor = GetActorWithName(GetWorld(), UTF8_TO_TCHAR(Request.actor().c_str()));
```
Comment (pete-tempo) [inline TempoWorld/Source/TempoWorld/Private/TempoActorControlServiceSubsystem.cpp:514]: Return early if actor not specified

File: TempoWorld/Source/TempoWorld/Private/TempoActorControlServiceSubsystem.cpp [hunk 15]
```diff
@@ -0,0 +1,834 @@
...
+			GetObjectProperties(ActorComponent, Response);
+		}
+	}
+
+	ResponseContinuation.ExecuteIfBound(Response, grpc::Status_OK);
+}
+
+void UTempoActorControlServiceSubsystem::GetComponentProperties(const GetComponentPropertiesRequest& Request, const TResponseDelegate<GetPropertiesResponse>& ResponseContinuation) const
+{
+	const AActor* Actor = GetActorWithName(GetWorld(), UTF8_TO_TCHAR(Request.actor().c_str()));
```
Comment (pete-tempo) [inline TempoWorld/Source/TempoWorld/Private/TempoActorControlServiceSubsystem.cpp:539]: Return early if actor or component not specified

File: TempoWorld/Source/TempoWorld/Private/TempoActorControlServiceSubsystem.cpp [hunk 16]
```diff
@@ -0,0 +1,834 @@
...
+
+template<>
+grpc::Status SetPropertyImpl<SetRotatorPropertyRequest>(const UWorld* World, const SetRotatorPropertyRequest& Request)
+{
+	FRotator Rotator;
+	Rotator.Roll = Request.r();
+	Rotator.Pitch = Request.p();
+	Rotator.Yaw = Request.y();
+
+	return SetStructPropertyImpl(World, Request, Rotator, TEXT("FRotator"));
```
Comment (pete-tempo) [inline TempoWorld/Source/TempoWorld/Private/TempoActorControlServiceSubsystem.cpp:756]: Probably should convert these from radians to degrees

File: TempoWorld/Source/TempoWorld/Public/ActorControl.proto (hunk 1 above)
Comment (pete-tempo) [reply TempoWorld/Source/TempoWorld/Public/ActorControl.proto:16]: Done, and also added to the response of finish spawning (the transform can also be adjusted on finish spawning).

File: TempoWorld/Source/TempoWorld/Private/TempoActorControlServiceSubsystem.cpp (hunk 12 above)
Comment (pete-tempo) [reply TempoWorld/Source/TempoWorld/Private/TempoActorControlServiceSubsystem.cpp:361]: There is no readme yet

**Review by gio-tempo — APPROVED** [review]

**Pull request merged**